ricecooker/utils/downloader.py

```diff
@@ -41,3 +54,180 @@ def read(path, loadjs=False, session=None, driver=None):
+                    filename=filename, middleware_callbacks=content_middleware)
+
+    def js_content_middleware(content, url, **kwargs):
+        if js_middleware:
+            content = js_middleware(content, url, **kwargs)
+
+        # Polyfill localStorage and document.cookie as iframes can't access
+        # them
+        return (content
+            .replace("localStorage", "_localStorage")
+            .replace('document.cookie.split', '"".split')
+            .replace('document.cookie', 'window._document_cookie'))
+
+    def css_node_filter(node):
+        return "stylesheet" in node["rel"]
+
+    def css_content_middleware(content, url, **kwargs):
+        if css_middleware:
+            content = css_middleware(content, url, **kwargs)
+
+        file_dir = os.path.dirname(urlparse(url).path)
+
+        # Download linked fonts and images
+        def repl(match):
+            src = match.group(1)
+            if src.startswith('//localhost'):
+                return 'url()'
+            # Don't download data: files
+            if src.startswith('data:'):
+                return match.group(0)
+            src_url = make_fully_qualified_url(os.path.join(file_dir, src), base_url)
+
+            if _is_blacklisted(src_url, url_blacklist):
+                print('        Skipping downloading blacklisted url', src_url)
+                return 'url()'
+
+            derived_filename = _derive_filename(src_url)
+            download_file(src_url, destination, request_fn=make_request,
+                    filename=derived_filename)
+            return 'url("%s")' % derived_filename
+
+        return _CSS_URL_RE.sub(repl, content)
+
+    # Download all linked static assets.
+    download_assets("img[src]", "src")  # Images
+    download_assets("link[href]", "href",
+            content_middleware=css_content_middleware,
+            node_filter=css_node_filter)  # CSS
+    download_assets("script[src]", "src",
+            content_middleware=js_content_middleware) # JS
+    download_assets("source[src]", "src") # Potentially audio
+    download_assets("source[srcset]", "srcset") # Potentially audio
+
+    # ... and also run the middleware on CSS/JS embedded in the page source to
+    # get linked files.
+    for node in doc.select('style'):
+        node.string = css_content_middleware(node.get_text(), url='')
+
+    for node in doc.select('script'):
+        if not node.attrs.get('src'):
+            node.string = js_content_middleware(node.get_text(), url='')
+
+    return doc
+
+
+def _is_blacklisted(url, url_blacklist):
+    return any((item in url.lower()) for item in url_blacklist)
+
+
+# TODO(davidhu): Use MD5 hash of URL (ideally file) instead.
+def _derive_filename(url):
+    name = os.path.basename(urlparse(url).path).replace('%', '_')
+    return ("%s.%s" % (uuid.uuid4().hex, name)).lower()
+
+
+def make_fully_qualified_url(url, base):
+    # TODO(davidhu): This is some hacky special-casing for 3asafeer ...
```

Comment:
I believe some of these can be handled by `urllib.parse.urljoin`, e.g., 
```
from urllib.parse import urljoin

base = 'http://somesite.com/subpath/file.html'

print(urljoin(base, '../image'))
http://somesite.com/image

print(urljoin(base, '../scripts'))
http://somesite.com/scripts

print(urljoin(base, '//some.full.path/just/no/scheme.html'))
http://some.full.path/just/no/scheme.html

print(urljoin(base, '/an/abs/path.html'))
http://somesite.com/an/abs/path.html

print(urljoin(base, 'a/rel/path.html'))
http://somesite.com/subpath/a/rel/path.html
```

Reply:
This is _awesome_, thanks for sharing!!